Refine exception stack API

c42f opened the pull request:
## Updated Description

This PR has been simplified with some parts merged separately (#30900, #30899). However, the `catch_stack` API has not been finalized. In the current version (as of 2019-02-06), I've got two things here:

* Rename `catch_stack()` -> `current_exceptions()` as this seems more descriptive.
* Introduce a new type `ExceptionStack` so that we can have a proper exception-and-backtrace specific type we can dispatch on. For example, this allows us to have a `show` implemented for exception and backtrace data which pretty prints the information in a much more readable way than the simple `Vector` of tuples that `catch_stack` returns right now.

## Original Description

Here's a few refinements to the API introduced in #28878. I'm not quite happy with that API for various reasons so I want to revisit it before 1.1 feature freeze (or unexport it before 1.1).

IMO this is a big improvement, but still a bit WIP so feedback would be great.

* Rename `catch_stack()` -> `current_exceptions()` as this seems far more transparent, and is in fact how the function is summarized in the docs. I wasn't happy with the previous name, it's got an analogy to `catch_backtrace` but other than that seemes too obscure. Alternative name proposals gladly welcome (the `current_exceptions(task)` variant is perhaps less natural).

* Introduce a new type `ExceptionStack` so that we have a proper exception-and-backtrace specific type we can dispatch on. This allows us to treat `showerror` as more of an implementation detail, and simply override  `show(::IO,::ExceptionStack)` in the usual way to get integrated pretty printing of exceptions and backtraces. I'm not quite certain about how this is a wrapper for a `Vector`; perhaps it should *be* an `AbstractVector`.

* For PARTR sanity, prevent reading of current exceptions of a concurrently running task, at least for now.

* Pretty printing of ExceptionStack in the REPL by overloading `display_error`. Could go into a separate PR if people want this simplified (or just review the commit 074e39b in isolation).

* Pretty printing of ExceptionStack via `show()`.  Examples follow:

```julia
try
    1 + [1,2]
catch exc
    try
        1÷0
    catch
        show(current_exceptions())
    end
end
```

![excstack_show](https://user-images.githubusercontent.com/601473/47918399-3a874780-def8-11e8-862b-a12c8ba27f1b.png)

Notice that by having proper `show` integration, we automatically get things like the following:
```julia
try
    1 + [1,2]
catch exc
    try
        1÷0
    catch
        @info "Ignoring an error" exception=current_exceptions()
    end
end
```

![excstack_logging](https://user-images.githubusercontent.com/601473/47918391-365b2a00-def8-11e8-911f-8f8465719fbb.png)

c42f commented:
Regarding `ExceptionStack`, we could alternatively model this as

```julia
struct ExceptionState
    exception
    backtrace
end

const ExceptionStack = Vector{ExceptionState}
```

which might make more or less sense. It's a little nicer because it doesn't introduce a container type for this one small thing. It's a little worse because it would assume that any `Vector{ExceptionState}` has the semantics of an exception stack rather than a simple array.

c42f commented:
Travis failure on linux32 is failure of a single test, due to #29923

The appveyor failure on 64 bit is really odd (a timeout? but with no error message or other indication of what went wrong?) I've restarted that and we'll see.

c42f commented:
Win64 CI problem has symptoms the same as #29603

StefanKarpinski commented:
Would it make sense to have each exception have `next::Union{Nothing,Exception}` field instead? That way you can ask for the current exception what, if any, the previous exception was.

c42f commented:
It does make sense and in some ways would be more natural and less disruptive. But the problem is user exception types: how do we ensure they actually have such a field? Basically, I don't think we can, at least not without the change being hugely breaking. We even allow people to throw things which aren't `<:Exception`! I tried to argue that this was a bad idea before but I'm not sure people were convinced (see https://github.com/JuliaLang/julia/issues/21886).

Given we don't have structural inheritance, that leaves us with composition:

```
struct ExceptionState
    exception
    backtrace
    next::Union{Nothing,ExceptionState}
end
```

Which I think would be quite a natural way of expressing this. But unfortunately no less breaking and somewhat less flexible to use than just doing what I've done in this PR, and shoving them all in a `Vector`. Thoughts?

StefanKarpinski commented:
Hmm. That's interesting: this may be the first compelling case I've encountered where allowing abstract types to have fields would be really useful—it would allow us to insert a `next` field into every subtype of `Exception`.

c42f commented:
By the way, my feeling was that there's no harm in providing a simple function-based interface to the exception stack system and seeing how that pans out.

Clearly it's not the end game and you'd like it to integrate better with the syntax. But the best syntax also depends on deeper design questions about how exceptions should be matched.

... having said that I didn't want to talk syntax, now I can't resist suggesting the following which popped into my head:

```julia
try
    # blah
catch es...  # equivalent to `es = current_exceptions()`
    # whatever
end
```

vchuravy commented on base/exports.jl:
```diff
@@ -679,7 +679,7 @@ export
 # errors
     backtrace,
     catch_backtrace,
-    catch_stack,
```
Probably we want a deprecation/alias for `catch_stack`?

c42f replied:
`catch_stack` has only existed for a few months during the current dev cycle, so I don't have any qualms about removing it entirely. In fact that's one of the more important points of this PR: to clean it up so that the API is in a better state for 1.1.

JeffBezanson commented on base/error.jl:
```diff
@@ -91,28 +91,34 @@ function catch_backtrace()
 uncaught exceptions.
 """
-function catch_stack(task=current_task(); include_bt=true)
+function current_exceptions(task=current_task(); include_bt=true)
```
Maybe `backtrace` instead of `include_bt`?

c42f replied:
Sure sounds good.

c42f replied:
Done.

JeffBezanson commented on base/error.jl:
```diff
@@ -91,28 +91,34 @@ function catch_backtrace()
     return _reformat_bt(bt[], bt2[])
 end
 
+struct ExceptionStack <: AbstractArray{Any,1}
```
This type doesn't seem necessary to me. Since showing errors goes through `display_error` anyway, having a type with a `show` method is a bit redundant (and does have some duplicated code here).

c42f replied:
Not that I love `ExceptionStack` as a type, it seems annoying and superfluous on the surface, and it's one of the main reasons I wanted some more eyes on this.

What I'd really like are suggestions about how to replace or improve it, while also having a way to pass these to logging macros and have them robustly recognized by the logger backend.

c42f replied:
It's true that there's a little duplication of code between `show` and `display_error`. Previously I tried making `display_error` dispatch to `show` and and adding some `IOContext` handling to turn on the simplifications we currently use in `display_error`. That turned out somewhat complex though so I punted in the current iteration and left it with to concrete implementations.

I could give it another go.

c42f commented:
I've rebased, fixed the minor conflict and renamed `include_bt`.

The remaining question in my mind is what to do about `ExceptionStack`. It's annoying to have a new type, but there's also reasons to want a type for this so that it can be `show()`n (and otherwise dispatched on) in contexts which are more general than REPL error reporting.

c42f commented:
Win64 issue appears to be #29603 again, restarted.

JeffBezanson commented:
My impression is that an exception or exception stack needs to be passed to a context that expects it --- it usually doesn't make sense to have a context that expects an arbitrary value, and then sometimes treats it as an error based on its type.

c42f commented:
Yes logging backends can have magic autodetection of "I think this vector looks structurally like it contains exceptions and backtrace pairs" and they can dig exception stacks out of the arbitrary list of user-supplied `key=value` pairs. I'm just somewhat uncomfortable having to re-infer this information when it was present in the first place.

I'm also not sure why `showerror` isn't just `show` with some special `IOContext` variable set, and this would be much easier to do if we had a special purpose type.

Putting those concerns aside in the interest of getting this done, would it be ok to define a type alias

```
const ExceptionStack = Array{NamedTuple{(:exception, :backtrace),<:Tuple}, 1}
```

and to remove the special `show` for `ExceptionStack` (people can use `showerror` if they like)? In this case we will also need special handling inside `ConsoleLogger` so that we can call `showerror` there rather than `show`.

c42f commented:
@JeffBezanson As food for thought I've added a commit c886fefd7214c0e24d9ea1453cc8c302eeceb363 which removes both `ExceptionStack` and the specialized `show`, but keeps a type alias. Because `show` was removed, I had to add something to `ConsoleLogger` so that exception stacks can be displayed in log messages in a useful form.

It's not too bad but on the whole it seems like a semantic step backward. On the upside it removes a small amount of code.

JeffBezanson commented:
> Yes logging backends can have magic autodetection of "I think this vector looks structurally like it contains exceptions and backtrace pairs" and they can dig exception stacks out of the arbitrary list of user-supplied `key=value` pairs. I'm just somewhat uncomfortable having to re-infer this information when it was present in the first place.

That's not at all what I meant to suggest. Looking at the overall type of the value, or the element type, or the types of objects contained in it are all variations of the same thing: examining the value itself to determine how to use it. I'm saying there needs to be some information outside the value itself to tell you what it represents. For example as @vtjnash mentioned `notify(cond, val, error=true)` treats `val` as an exception. We don't do that based on the type of `val`. Similarly a special key could be used in logging e.g. `exception=x`.

c42f commented:
In the spirit of seeing how `LoadError`/`InitError` can be made more consistent with the new system, I've changed them from using `jl_rethrow_other` to `jl_throw`. With this PR's REPL changes, these would be reported as follows:

![2018-12-07-190857_784x460_scrot](https://user-images.githubusercontent.com/601473/49638328-97c06c80-fa53-11e8-9975-361e9352fcac.png)

The previous exception is still eagerly captured and stored in `LoadError`. Not sure how that could (or should) be changed without breaking compatibility.

JeffBezanson commented:
Yes, I agree LoadError (and InitError) saving the last exception is now redundant; nice! I would say keep everything the same except the printing of LoadError, which now doesn't need to show its inner exception. In 2.0 we can remove those fields.

vtjnash commented:
Unrelated to this PR, but just from looking at that screenshot, we should eventually hide all frames (but one) that are printed in a earlier (in printing order) list

JeffBezanson commented:
Also the simplest thing to do here is just unexport `catch_stack` for now.

StefanKarpinski commented:
> Also the simplest thing to do here is just unexport `catch_stack` for now.

Need to make sure no packages are using `catch_stack` anywhere.

c42f commented:
Looks like this will have to miss 1.1. I'll make a PR to the release-1.1 branch to unexport `catch_stack` for now and mark it as experimental. @StefanKarpinski how would I reliably find such packages, and isn't it their problem if they use features which have only been on the dev branch?

StefanKarpinski commented:
Oh, if it's only on the dev branch then it's fine to delete.

c42f commented:
For now `catch_stack` is unexported and marked experimental (#30316 merged to master, pending backport).

The REPL display-related changes from here could be made non-breaking and backported to 1.1 release branch if desired so that people at least get the benefit of seeing the full exception stack even though the API isn't finalized.

c42f commented:
This PR has become unwieldy. I'm going to pull it apart into a few pieces so we can get the straightforward and useful pieces into 1.2 independently of deciding on the correct API for `catch_stack`.

c42f commented:
Ok, I've merged all of the REPL-related `display_error` changes separately. This PR is now purely about resolving the naming of `catch_stack` and the appropriate return type.

Regarding whether we need an `ExceptionStack` type — I still think something like it is useful. It's not about `ExceptionStack` being an error (which is context-dependent as discussed above), but about code being able to recognize the stack as a structured set of exceptions, rather than a `Vector` which *happens* to contain some exceptions and backtrace-like things.

This is most obviously useful for `show`; consider which of the following options is more legible (version 1.1.0 and the latest version of this PR, respectively):

![stack_printing](https://user-images.githubusercontent.com/601473/52316167-f24f5800-2a05-11e9-9293-aefec51f87bd.png)

Keno commented:
Triage tag was outdated. If there's something left to triage, re-apply the tag and specify the issue.

vtjnash reviewed (approved):
LGTM, conditional on addressing the needs-tests tag (and needs a rebase).

vtjnash commented:
bump. Are you able to rebase?

c42f committed:
Exception stack API refinements

* Rename the non-exported `catch_stack()` to the more descriptive name
  `current_exceptions()`. Keep the old name available but deprecated.

* Introduce an ExceptionStack as the return type for the function, which
  (as an AbstractVector) is API-compatible with the previous type
  returned by `catch_stack()`

Having ExceptionStack gives us a place to integrate exception printing
in a natural way. In the same way this should be useful for dispatch in
other areas of the ecosystem which want to dispatch on exception stacks.

c42f commented:
Wow, this PR has languished for a long time :-(

I've found some time to rebase this and update the tests.

I'll plan to merge it tomorrow, provided CI passes.
